vm.c

```diff
@@ -164,3 +172,155 @@ void vm_run(vm_env *env)
+    char *ptr;
+    char *end;
+    int i = 0;
+
+    memset(&env->cpool[0], 0, sizeof(env->cpool));
+
+    src = (vm_value *)mem;
+    dst = &env->cpool[0];
+    end = mem + sz;
+
+    while ((char *)src < end)
+    {
+        memcpy(dst, src, sizeof(*dst));
+
+        if (src->type == STR) {
+            ptr = &mem[(int)src->value.vstr];
+
+            dst->value.vstr = strdup(ptr);
+
+            if (ptr < end)
+                end = ptr;
+        }
+
+        dst++;
+        src++;
+        i++;
+    }
+
+    return env->cpool_count = i++;
+}
+
+static int vm_cpool_seg_deflate(vm_env *env, char **mem, int *sz)
+{
+    vm_value *src, *dst;
+    char *ptr;
+
+    *sz = env->cpool_count * sizeof(vm_value);
+
+    for (int i = 0; i < env->cpool_count; i++)
+        if (env->cpool[i].type == STR)
+            *sz += strlen(env->cpool[i].value.vstr) + 1;
+
+    *mem = malloc(*sz);
+    memcpy(*mem, &env->cpool[0], env->cpool_count * sizeof(vm_value));
+
+    dst = (vm_value *)*mem;
+    src = (vm_value *)&env->cpool[0];
+    ptr = &((*mem)[ env->cpool_count * sizeof(vm_value) ]);
+
+    for (int i = 0; i < env->cpool_count; i++) {
+        if (src[i].type == STR) {
+            dst[i].value.vint = ptr - *mem; // use vint to avoid casting
+            strcpy(ptr, src[i].value.vstr);
+            ptr += strlen(src[i].value.vstr) + 1;
+        }
+    }
+
+    return *sz;
+}
+
+static int vm_temps_seg_inflate(vm_env *env, char *mem, int sz)
+{
+    memset(&env->temps[0], 0, sizeof(env->temps));
+    memcpy(&env->temps[0], mem, sz);
+    return env->temps_count = sz / sizeof(vm_value);
+}
+
+static int vm_temps_seg_deflate(vm_env *env, char **mem, int *sz)
+{
+    *sz = env->temps_count * sizeof(vm_value);
+    *mem = malloc(*sz);
+
+    memcpy(*mem, (char *) &env->temps[0], *sz);
+
+    return *sz;
+}
+
+unsigned short vm_get_seg_info(vm_env *env, vm_seg_info **seg_info)
+{
+    vm_seg_info *p = *seg_info = zalloc(sizeof(*p));
+
+    vm_insts_seg_deflate(env, &p->mem, &p->sz);
+    p->next = zalloc(sizeof(*p));
+    p = p->next;
+
+    vm_cpool_seg_deflate(env, &p->mem, &p->sz);
+    p->next = zalloc(sizeof(*p));
+    p = p->next;
+
+    vm_temps_seg_deflate(env, &p->mem, &p->sz);
+    p->next = NULL;
+
+    return 3;
+}
+
+void vm_free_seg_info(vm_seg_info *seg_info)
+{
+    vm_seg_info *p;
+
+    while (seg_info) {
+        p = seg_info->next;
+        free(seg_info->mem);
+        free(seg_info);
+        seg_info = p;
+    }
+}
+
+int vm_set_seg_info(vm_env *env, const vm_seg_info *seg_info)
+{
+    int i = 0;
+
+    while (seg_info) {
+        switch (i) {
+        case 0:
+            vm_insts_seg_inflate(env, seg_info->mem, seg_info->sz);
+            break;
+        case 1:
+            vm_cpool_seg_inflate(env, seg_info->mem, seg_info->sz);
+            break;
+        case 2:
+            vm_temps_seg_inflate(env, seg_info->mem, seg_info->sz);
+            break;
+        default:
+            break;
+        }
+        i++;
```

Comment:
I didn't get the reason why you tended to look up incremented `i`. Can you explain?